File: src/R/systems/validate_phase12c_freeze.R
#!/usr/bin/env Rscript
# Independent base-R freeze-boundary validation for accepted Phase 12C.
args <- commandArgs(trailingOnly=TRUE)
root <- if (length(args)) normalizePath(args[[1]], mustWork=TRUE) else normalizePath(".", mustWork=TRUE)
reports <- file.path(root, "reports")

stop_if <- function(condition, message) if (!isTRUE(condition)) stop(message, call.=FALSE)
json_text <- function(path) paste(readLines(path, warn=FALSE, encoding="UTF-8"), collapse=" ")
sha256_file <- function(path) {
  cmd <- Sys.which("sha256sum")
  if (nzchar(cmd)) {
    out <- system2(cmd, path, stdout=TRUE, stderr=TRUE)
    stop_if(length(out) >= 1L, paste("sha256sum failed", path))
    return(tolower(gsub("[^0-9a-f]", "", strsplit(trimws(out[[1]]), "[[:space:]]+")[[1]][1])))
  }
  certutil <- Sys.which("certutil")
  stop_if(nzchar(certutil), "Neither sha256sum nor certutil is available")
  out <- system2(certutil, c("-hashfile", path, "SHA256"), stdout=TRUE, stderr=TRUE)
  hits <- tolower(gsub("[[:space:]]+", "", out)); hits <- hits[grepl("^[0-9a-f]{64}$", hits)]
  stop_if(length(hits) >= 1L, paste("certutil failed", path)); hits[[1]]
}
portable_match <- function(path, expected) {
  if (sha256_file(path) == tolower(expected)) return(TRUE)
  ext <- tolower(tools::file_ext(path))
  if (!(ext %in% c("csv", "json", "md", "txt", "yml", "yaml", "svg", "py", "r"))) return(FALSE)
  raw <- readBin(path, "raw", n=as.numeric(file.info(path)$size))
  canonical <- charToRaw(gsub("\r\n?", "\n", rawToChar(raw), perl=TRUE))
  crlf <- charToRaw(gsub("\n", "\r\n", rawToChar(canonical), fixed=TRUE))
  lf_tmp <- tempfile(); crlf_tmp <- tempfile()
  on.exit(unlink(c(lf_tmp, crlf_tmp)), add=TRUE)
  writeBin(canonical, lf_tmp); writeBin(crlf, crlf_tmp)
  tolower(expected) %in% c(sha256_file(lf_tmp), sha256_file(crlf_tmp))
}
manifest_artifacts <- function(path) {
  lines <- readLines(path, warn=FALSE, encoding="UTF-8")
  collection <- if (any(grepl('"artifacts"[[:space:]]*:[[:space:]]*\\{', lines, perl=TRUE))) "artifacts" else "files"
  start <- which(grepl(paste0('"', collection, '"'), lines, fixed=TRUE) & grepl("{", lines, fixed=TRUE))[1]
  stop_if(!is.na(start), paste("missing manifest collection", path))
  rel <- character(); hashes <- character(); current <- NA_character_; current_hash <- NA_character_
  append_current <- function() {
    if (!is.na(current) && !is.na(current_hash)) { rel <<- c(rel, current); hashes <<- c(hashes, current_hash) }
    current <<- NA_character_; current_hash <<- NA_character_
  }
  if (start + 1L <= length(lines)) for (line in lines[(start + 1L):length(lines)]) {
    if (grepl("^  \\},?[[:space:]]*$", line)) { append_current(); break }
    direct <- regexec('^[[:space:]]*"([^"]+/[^"]+)"[[:space:]]*:[[:space:]]*"([0-9a-f]{64})"', line, perl=TRUE)
    hit <- regmatches(line, direct)[[1]]
    if (length(hit)==3L) { append_current(); rel <- c(rel, hit[[2]]); hashes <- c(hashes, hit[[3]]); next }
    nested <- regexec('^[[:space:]]*"([^"]+/[^"]+)"[[:space:]]*:[[:space:]]*\\{', line, perl=TRUE)
    hit <- regmatches(line, nested)[[1]]
    if (length(hit)==2L) { append_current(); current <- hit[[2]] }
    h <- regexec('"sha256"[[:space:]]*:[[:space:]]*"([0-9a-f]{64})"', line, perl=TRUE)
    hh <- regmatches(line, h)[[1]]
    if (length(hh)==2L && !is.na(current)) current_hash <- hh[[2]]
  }
  append_current()
  data.frame(rel=rel, sha256=hashes, stringsAsFactors=FALSE)
}
check_manifest <- function(path, expected_n, phase, baseline) {
  stop_if(file.exists(path), paste("missing freeze manifest", path))
  txt <- json_text(path)
  stop_if(grepl('"status"[[:space:]]*:[[:space:]]*"ACCEPTED / FROZEN"', txt, perl=TRUE), paste("status", path))
  stop_if(grepl(paste0('"accepted_phase"[[:space:]]*:[[:space:]]*"', phase, '"'), txt, perl=TRUE), paste("phase", path))
  stop_if(grepl(baseline, txt, fixed=TRUE), paste("baseline", path))
  x <- manifest_artifacts(path)
  stop_if(nrow(x) == expected_n, paste("freeze artifact count", path))
  for (i in seq_len(nrow(x))) {
    artifact <- file.path(root, x$rel[[i]])
    stop_if(file.exists(artifact) && portable_match(artifact, x$sha256[[i]]), paste("freeze hash", x$rel[[i]]))
  }
  nrow(x)
}

A <- file.path(reports, "phase12a_vector_ecology_freeze_manifest.json")
B <- file.path(reports, "phase12b_vector_environment_human_dependencies_freeze_manifest.json")
C <- file.path(reports, "phase12c_vector_ecology_futures_freeze_manifest.json")
W <- file.path(reports, "vector_ecology_future_manifest.json")
K <- file.path(reports, "vector_ecology_future_artifact_check.json")
stop_if(check_manifest(A, 17L, "12A", "Vector Ecology Baseline, 2026") == 17L, "Phase 12A")
stop_if(check_manifest(B, 15L, "12B", "Vector / Environment / Human-System Dependencies, 2026") == 15L, "Phase 12B")
stop_if(check_manifest(C, 28L, "12C", "Vector Ecology Futures, 2050 / 2075") == 28L, "Phase 12C")

working_text <- json_text(W)
stop_if(grepl('"status"[[:space:]]*:[[:space:]]*"implemented_validated_pending_sol_acceptance"', working_text, perl=TRUE), "working manifest status")
working_artifacts <- manifest_artifacts(W)
stop_if(nrow(working_artifacts) == 25L, "working manifest artifact count")
for (i in seq_len(nrow(working_artifacts))) stop_if(portable_match(file.path(root, working_artifacts$rel[[i]]), working_artifacts$sha256[[i]]), paste("working hash", working_artifacts$rel[[i]]))

check_text <- json_text(K)
for (term in c('"status": "passed"', '"scenario_assumptions": 36', '"vector_states": 36', '"habitat_states": 48', '"surveillance_states": 30', '"dependency_states": 168', '"uncertainty_states": 60', '"scenario_sources": 28', '"comparison_rows": 6', '"prior_phase1_11_manifest_entries": 298', '"prior_phase1_11_unique_protected_artifacts": 293', '"phase12a_12b_freeze_integrity": true', '"phase12a_12b_immutable": true', '"active_holds_preserved": true', '"independent_review_passed": true')) stop_if(grepl(term, check_text, fixed=TRUE), paste("artifact check", term))

final_text <- json_text(C)
for (term in c('"phase12a_12b_immutable": true', '"phase1_11_immutable": true', '"phase13_implemented": false', 'reports/phase12a_vector_ecology_freeze_manifest.json', 'reports/phase12b_vector_environment_human_dependencies_freeze_manifest.json', 'reports/vector_ecology_future_manifest.json', '"great_black_swamp": "C — HOLD / noncanonical"', '"toledo_intake_coordinate_discrepancy": "UNRESOLVED"')) stop_if(grepl(term, final_text, fixed=TRUE), paste("final boundary", term))

review_text <- tolower(json_text(file.path(reports, "vector_ecology_future_independent_review.md")))
for (term in c('"passed": true', '"security_concerns": []', '"logic_errors": []', '"provenance_errors": []', '"ecological_errors": []', '"scenario_boundary_errors": []', '"spatial_scale_errors": []', '"health_boundary_errors": []')) stop_if(grepl(term, review_text, fixed=TRUE), paste("review", term))
stop_if(file.exists(file.path(reports, "vector_ecology_future_independent_review_initial.md")), "initial review preserved")

prior_files <- list.files(reports, pattern="^phase.*_freeze_manifest[.]json$", full.names=TRUE)
prior_files <- prior_files[!grepl("^phase12", basename(prior_files))]
protected <- character(); expected_hashes <- character(); prior_entries <- 0L
for (path in sort(prior_files)) {
  x <- manifest_artifacts(path)
  for (i in seq_len(nrow(x))) {
    rel <- x$rel[[i]]; expected_hash <- x$sha256[[i]]
    if (rel %in% names(expected_hashes)) stop_if(expected_hashes[[rel]] == expected_hash, paste("prior duplicate hash", rel))
    expected_hashes[rel] <- expected_hash
    stop_if(file.exists(file.path(root, rel)) && portable_match(file.path(root, rel), expected_hash), paste("prior hash", rel))
    protected <- c(protected, rel); prior_entries <- prior_entries + 1L
  }
}
stop_if(prior_entries == 298L && length(unique(protected)) == 293L, paste("prior inventory", prior_entries, length(unique(protected))))

cat('{\n  "status": "passed",\n  "phase": "12C",\n  "freeze_manifest": "reports/phase12c_vector_ecology_futures_freeze_manifest.json",\n  "freeze_artifacts": 28,\n  "working_manifest_artifacts": 25,\n  "phase12a_artifacts": 17,\n  "phase12b_artifacts": 15,\n  "prior_phase1_11_manifest_entries": 298,\n  "prior_phase1_11_unique_protected_artifacts": 293,\n  "phase12a_12b_immutable": true,\n  "review_passed_blocking_arrays_empty": true,\n  "active_holds_preserved": true,\n  "phase13_implemented": false\n}\n')

File: src/python/systems/validate_phase12c_freeze.py
#!/usr/bin/env python3
"""Independent freeze-boundary validation for the accepted Phase 12C package."""
from __future__ import annotations

import argparse
import hashlib
import json
import re
from pathlib import Path
from typing import Any

EXPECTED_COUNTS = {
    "scenario_assumptions": 36,
    "vector_states": 36,
    "habitat_states": 48,
    "surveillance_states": 30,
    "dependency_states": 168,
    "uncertainty_states": 60,
    "scenario_sources": 28,
    "comparison_rows": 6,
}
EXPECTED_PRIOR_ENTRIES = 298
EXPECTED_PRIOR_UNIQUE = 293
PHASE12A_MANIFEST = "reports/phase12a_vector_ecology_freeze_manifest.json"
PHASE12B_MANIFEST = "reports/phase12b_vector_environment_human_dependencies_freeze_manifest.json"
PHASE12C_MANIFEST = "reports/phase12c_vector_ecology_futures_freeze_manifest.json"
WORKING_MANIFEST = "reports/vector_ecology_future_manifest.json"
ARTIFACT_CHECK = "reports/vector_ecology_future_artifact_check.json"

TEXT_SUFFIXES = {
    ".csv", ".json", ".md", ".py", ".r", ".R", ".txt", ".yaml", ".yml", ".toml",
    ".html", ".css", ".js", ".ts", ".tsx", ".jsx", ".svg", ".xml",
}


def load_json(root: Path, rel: str) -> dict[str, Any]:
    with (root / rel).open("r", encoding="utf-8") as handle:
        return json.load(handle)


def raw_digest(data: bytes) -> str:
    return hashlib.sha256(data).hexdigest()


def normalized_digest(data: bytes) -> str:
    return raw_digest(data.replace(b"\r\n", b"\n").replace(b"\r", b"\n"))


def verify_hash(root: Path, rel: str, expected: dict[str, Any] | str) -> str:
    path = root / rel
    assert path.is_file(), f"missing protected artifact: {rel}"
    data = path.read_bytes()
    if isinstance(expected, str):
        expected_hash = expected
        expected_bytes = None
    else:
        expected_hash = expected.get("sha256")
        expected_bytes = expected.get("bytes")
    raw_hash = raw_digest(data)
    if raw_hash == expected_hash and (expected_bytes is None or len(data) == expected_bytes):
        return "exact"
    if path.suffix in TEXT_SUFFIXES and normalized_digest(data) == expected_hash:
        normalized_len = len(data.replace(b"\r\n", b"\n").replace(b"\r", b"\n"))
        assert expected_bytes in (None, len(data), normalized_len), f"byte mismatch: {rel}"
        return "newline-normalized"
    raise AssertionError(f"hash/byte mismatch: {rel}")


def verify_artifacts(root: Path, manifest: dict[str, Any], label: str) -> dict[str, int]:
    exact = 0
    normalized = 0
    artifacts = manifest.get("artifacts", {})
    assert artifacts, f"{label}: no artifacts"
    for rel, expected in artifacts.items():
        mode = verify_hash(root, rel, expected)
        if mode == "exact":
            exact += 1
        else:
            normalized += 1
    return {"entries": len(artifacts), "exact": exact, "newline_normalized": normalized}


def prior_manifests(root: Path) -> list[Path]:
    found: list[Path] = []
    for path in sorted((root / "reports").glob("*freeze_manifest.json")):
        if path.name not in {Path(PHASE12A_MANIFEST).name, Path(PHASE12B_MANIFEST).name, Path(PHASE12C_MANIFEST).name}:
            found.append(path)
    return found


def verify_prior_system(root: Path) -> dict[str, int]:
    manifests = prior_manifests(root)
    entries: list[str] = []
    normalized = 0
    for path in manifests:
        data = json.loads(path.read_text(encoding="utf-8"))
        for rel, expected in data.get("artifacts", data.get("files", {})).items():
            entries.append(rel)
            if verify_hash(root, rel, expected) == "newline-normalized":
                normalized += 1
    unique = len(set(entries))
    assert len(entries) == EXPECTED_PRIOR_ENTRIES, len(entries)
    assert unique == EXPECTED_PRIOR_UNIQUE, unique
    return {
        "manifest_count": len(manifests),
        "entries": len(entries),
        "unique": unique,
        "newline_normalized": normalized,
    }


def review_verdict(root: Path) -> dict[str, Any]:
    text = (root / "reports/vector_ecology_future_independent_review.md").read_text(encoding="utf-8")
    match = re.search(r"```json\s*(\{.*?\})\s*```", text, flags=re.S)
    assert match, "final Phase 12C review JSON block missing"
    verdict = json.loads(match.group(1))
    assert verdict.get("passed") is True
    blocking = (
        "security_concerns", "logic_errors", "provenance_errors", "ecological_errors",
        "scenario_boundary_errors", "spatial_scale_errors", "health_boundary_errors",
    )
    for key in blocking:
        assert verdict.get(key) == [], key
    assert (root / "reports/vector_ecology_future_independent_review_initial.md").is_file()
    return {"passed": True, "blocking_arrays_empty": True, "initial_review_preserved": True}


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("root", nargs="?", default=None, help="repository root")
    args = parser.parse_args()
    root = Path(args.root).resolve() if args.root else Path(__file__).resolve().parents[3]

    freeze = load_json(root, PHASE12C_MANIFEST)
    assert freeze["accepted_phase"] == "12C"
    assert freeze["status"] == "ACCEPTED / FROZEN"
    assert freeze["protected_inputs"] == [PHASE12A_MANIFEST, PHASE12B_MANIFEST, WORKING_MANIFEST]
    assert freeze["phase12a_12b_immutable"] is True
    assert freeze["phase1_11_immutable"] is True
    assert freeze["phase13_implemented"] is False
    assert freeze["active_holds_preserved"] == {
        "great_black_swamp": "C — HOLD / noncanonical",
        "toledo_intake_coordinate_discrepancy": "UNRESOLVED",
    }
    assert freeze["counts"] == EXPECTED_COUNTS
    freeze_result = verify_artifacts(root, freeze, "Phase 12C final freeze manifest")

    working = load_json(root, WORKING_MANIFEST)
    assert working["status"] == "implemented_validated_pending_sol_acceptance"
    assert working["counts"] == EXPECTED_COUNTS
    working_result = verify_artifacts(root, working, "Phase 12C working manifest")

    artifact_check = load_json(root, ARTIFACT_CHECK)
    assert artifact_check["status"] == "passed"
    for key, expected in EXPECTED_COUNTS.items():
        assert artifact_check[key] == expected, key
    assert artifact_check["prior_phase1_11_manifest_entries"] == EXPECTED_PRIOR_ENTRIES
    assert artifact_check["prior_phase1_11_unique_protected_artifacts"] == EXPECTED_PRIOR_UNIQUE
    assert artifact_check["phase12a_12b_freeze_integrity"] is True
    assert artifact_check["phase12a_12b_immutable"] is True
    assert artifact_check["active_holds_preserved"] is True
    assert artifact_check["independent_review_passed"] is True

    phase12a = load_json(root, PHASE12A_MANIFEST)
    phase12b = load_json(root, PHASE12B_MANIFEST)
    assert phase12a["status"] == "ACCEPTED / FROZEN"
    assert phase12b["status"] == "ACCEPTED / FROZEN"
    a_result = verify_artifacts(root, phase12a, "Phase 12A final freeze manifest")
    b_result = verify_artifacts(root, phase12b, "Phase 12B final freeze manifest")
    assert phase12b["phase12a_freeze_manifest"]["path"] == PHASE12A_MANIFEST
    assert verify_hash(root, PHASE12A_MANIFEST, phase12b["phase12a_freeze_manifest"]) in {"exact", "newline-normalized"}

    prior_result = verify_prior_system(root)
    review = review_verdict(root)
    forbidden = [p for p in freeze["artifacts"] if "phase13" in p.lower()]
    assert not forbidden

    result = {
        "status": "passed",
        "phase": "12C",
        "freeze_manifest": {"path": PHASE12C_MANIFEST, **freeze_result},
        "working_manifest": working_result,
        "phase12a": a_result,
        "phase12b": b_result,
        "prior_phase1_11": prior_result,
        "review": review,
        "counts": EXPECTED_COUNTS,
        "phase12a_12b_immutable": True,
        "active_holds_preserved": True,
        "phase13_implemented": False,
    }
    print(json.dumps(result, indent=2, ensure_ascii=False))


if __name__ == "__main__":
    main()
